mp4parse_capi/src/lib.rs

```diff
@@ -615,6 +640,149 @@ pub unsafe extern fn mp4parse_get_track_video_info(parser: *mut mp4parse_parser,
+            index_table.insert(track_id, v);
+            return MP4PARSE_OK;
+        },
+        _ => {},
+    }
+
+    MP4PARSE_ERROR_INVALID
+}
+
+fn create_sample_table(track: &Track) -> Option<Vec<mp4parser_indice>> {
+    let timescale = match track.timescale {
+        Some(t) => t,
+        _ => return None,
+    };
+
+    let (stsc, stco, stsz, stss, stts) =
+        match (&track.stsc, &track.stco, &track.stsz, &track.stss, &track.stts) {
+            (&Some(ref a), &Some(ref b), &Some(ref c), &Some(ref d), &Some(ref e)) => (a, b, c, d, e),
+            _ => return None,
+        };
+
+    // According to spec, no sync table means every sample is sync sample.
+    let has_sync_table = match track.stss {
+        Some(_) => true,
+        _ => false,
+    };
+
+    // Create sample table and calculate sample offset.
+    let mut sample_table = Vec::new();
+    let mut sample_size_iter = stsz.sample_sizes.iter();
+    let mut start_chunk = 0;
+    let mut end_chunk = 0;
+    let ret = stsc.samples.iter()
+        .map(|sample_chunk| {
+                // Convert the compact table to chunk range and samples per trunk.
+                start_chunk = end_chunk;
+                end_chunk += sample_chunk.first_chunk;
+                ((start_chunk .. end_chunk), sample_chunk.samples_per_chunk)
+            })
+        .all(|chunks| {
+                // Chunks in the chunk range have the same sample number.
+                let chunks_range = chunks.0;
+                let samples_per_chunk = chunks.1;
+                for chunk_id in chunks_range {
+                    let mut cur_position: u64 = stco.offsets[chunk_id as usize];
+                    // Create sample according to sample number of each trunk and calculate
+                    // sample offset.
+                    for _ in 0 .. samples_per_chunk {
+                        let start_offset = cur_position;
+                        let end_offset = match (stsz.sample_size, sample_size_iter.next()) {
+                            (_, Some(t)) => start_offset + *t as u64,
+                            (t, _) if t > 0 => start_offset + t as u64,
+                            _ => 0,
+                        };
+                        if end_offset == 0 {
+                            return false;
+                        }
+                        cur_position = end_offset;
+
+                        sample_table.push(
+                            mp4parser_indice {
+                                start_offset: start_offset,
+                                end_offset: end_offset,
+                                start_composition: 0,
+                                end_composition: 0,
+                                start_decode: 0,
+                                sync: !has_sync_table,
+                            }
+                        );
+                    }
+                }
+            true
+        });
+    if !ret {
+        return None;
```

Comment:
Since this check is so far from the declaration, please use a more descriptive variable name than `ret`. Maybe `have_table_entries` or `table_valid`, something like that, to be clear why you can't just check `sample_table.is_empty()`.